### PR TITLE
Proposal to return T instead or Response(of T)

### PR DESCRIPTION
**Pull request by johanste**
PR to use for discussion...

### CODE REVIEW

File: docs/design/python/response_of_t_or_t.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+##### Cons
+
+* Whenever there isn't a 1:1 mapping between client method call and HTTP request, it is less clear exactly *which* response the headers belong to.
+    * Ex. Paging and polling operations
+* The set of headers exposed by a service in their responses is not easy to intuit from the method signature.
+* Header values are untyped, leaving it to the caller to deserialize the value. Additionally, a caller must handle the presence of multiple headers for a response.
+
+### Methods return T
+In this model, we model T independently of how information was received (header or body). Headers related to the returned entity are modeled on T itself. Headers related to the *request* (e.g. request id, request processing cost) are assumed to primarily be of interest to logging and telemetry and thus consumed/"exposed" by logging and tracing policies. The full response is available in failures as a property in the raised exception.
```
Comment (KrzysztofCwalina) [inline docs/design/python/response_of_t_or_t.md]: I don't think policies "expose" the rest of the headers. It's impossible to correlate between client call and policy picking up a response. The reason is that pipeline is shared between many possibly concurrent client calls. And so the main cons of this approach is that it does not really allow users to access all headers.

File: docs/design/python/response_of_t_or_t.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+```
+Status: 200 NotFound
+x-ms-request-id: 14
+x-ms-session-token: xafg2
+```
+Python code: 
+```python
+>>> client = ExampleClient(...)
+>>> client.get_thing(8)
+>>> print(client.session_token)
```
Comment (KrzysztofCwalina) [inline docs/design/python/response_of_t_or_t.md]: How does it work if multiple threads use one shared client?

File: docs/design/python/response_of_t_or_t.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+```python
+>>> def my_deserializer(raw_response):
+>>>     wrapped_result = raw_response.deserialize()
+>>>     setattr(wrapped_result, 'request_id', raw_response.headers['x-ms-request-id'])
+>>>     return wrapped_result
+>>> 
+>>> thing = client.get_thing(2, response_handler=my_deserializer)
+>>> print(thing.request_id)
+15
+```
```
Comment (KrzysztofCwalina) [inline docs/design/python/response_of_t_or_t.md]: It seems like none of the samples let users read an arbitrary header from the response. I thought we wanted to make the common case easy but getting an arbitrary header possible.

Comment (johanste) [reply docs/design/python/response_of_t_or_t.md]: Replace `'x-ms-request-id'` with any arbitrary header.

File: docs/design/python/response_of_t_or_t.md (hunk 2 above)
Comment (johanste) [reply docs/design/python/response_of_t_or_t.md]: Consider the canonical scenario 'what is my remaining call quota' as a "connection based" piece of metadata returned. And you have multiple threads racing to get the value in the presence of in-flight requests. Pragmatically, last write wins. If you have semantic understanding of the value, you may ensure that you don't even expose out-of-order writes. But given that these values are typically server generated, you cannot be sure that the client observed the header values in a different order than the service included them in response. 

So, we only need to care about not having torn reads of the value.

File: docs/design/python/response_of_t_or_t.md (hunk 1 above)
Comment (johanste) [reply docs/design/python/response_of_t_or_t.md]: You are correct. I'll update the phrasing. I do have an implicit assumption that one can pass in a callback on a per-request basis.

File: docs/design/python/response_of_t_or_t.md (hunk 3 above)
Comment (KrzysztofCwalina) [reply docs/design/python/response_of_t_or_t.md]: I was not super clear. I added the comment to the end of the samples, but it's a general comment about not having an ability to read an arbitrary header unless it's added to the model type. It's related to the issue of not having per call callbacks.

**Comment by adrianhall** [comment]
## Architecture Board Discussion

**Comments**
* (BT) In the headers section, can you call it session?  
    * (JS) Some of them are per-account
* (KC/JS) Response<T> carries the response stream all the way up the chain
* (JR) There is an issue around what happens when the connection goes down during read (deserialization happens in the pipeline, but could be postponed to after the pipeline)
    * (JS) This is an orthogonal problem that will be there irrespective of the choice
* (KC) The biggest issue with Response<T> is it makes the signature of the API looks really busy
* (AH) There is a use case that suggests the status code needs to be available, but doesn't need to be in the main flow
* (JR) In the case of returning T, you don't have a way to get the status code.
* (AH) Can you create a CachingPolicy?  (Maybe, yes in Node, no in .NET)
* (BT) I can justify why caching SHOULD go in the pipeline and why the pipeline CAN be complex.  (ServiceWorker) - KC disagrees
* (KC) We are concerned about differences in success status codes, but headers are important

**Options (really)**:
1. You can't
2. Response<T>
3. Lift to properties of T (T.headers, etc.)
3. Response property on T
5. Optional callback (in \*\*kwargs) that takes response
6. GetMetadataFor(T)
7. out param via mutating options object
8. ThreadContext
9. Message as first-class entity

**Open Issues from discussion**:
* Pushing streams up the stack to the caller / consumer of the response (KC / JR / JS)

**Agreed Principles**:
* The HTTP response (status code and headers) is available, but not in your face

**Action Items**:
* (Language Architects) Determine what is required to support returning T
* (Language Architects) Design a UX study to see whether Response<T> is an issue

Re-review on 4/1

**Pull request closed without merging**